### PR TITLE
add a Justfile for reproducing CI locally, use it for CI

### PR DESCRIPTION
**Pull request by Ekleog-NEAR**

### CODE REVIEW

**Comment by Ekleog-NEAR** [comment]
I failed on my promise to have a PR with a Justfile ready yesterday, but I did do it today, so here it is! :)

File: Justfile [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+test: test-ci test-extra
+
+# only the tests that are exactly the same as the ones in CI
+test-ci: (nextest "stable") (nextest "nightly") check-style
+
+# tests that are as close to CI as possible, but not exactly the same code
+test-extra: check-lychee
+
+# all cargo tests, TYPE is "stable" or "nightly"
+nextest TYPE: (nextest-unit TYPE) (nextest-integration TYPE)
```
Comment (nagisa) [inline Justfile:22]: It is important that we allow passing filters and other arbitrary flags through to `nextest` as well… Local development will usually not run the full test suite every time, rather filtering for a subset of tests they are interested in with flags like `-p` and just plain test names.

File: Justfile [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+# cargo unit tests, TYPE is "stable" or "nightly"
+nextest-unit TYPE:
+    cargo nextest run \
+        --locked \
+        --workspace \
+        --exclude integration-tests \
+        --cargo-profile dev-release \
+        --profile ci \
```
Comment (nagisa) [inline Justfile:31]: Local development does not want `--profile ci`. It largely turns on retrying the tests, which goes in opposition with the goal that people would notice them during local development and perhaps consider fixing them.

File: Justfile [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+[macos]
+nextest-integration TYPE:
+    @echo "Nextest integration tests are currently disabled on macos!"
+
+# generate a codecov report for RULE
+codecov RULE:
+    #!{{ bash }}
+    set -euxo pipefail
+    source <(cargo llvm-cov show-env --export-prefix)
+    just {{ RULE }}
```
Comment (nagisa) [inline Justfile:57]:
```suggestion
    just_executable() {{ RULE }}
```

Unfortunate that it is necessary to recursively invoke `just` with a use-case as basic as this though… as it no longer provides the strong guarantees before starting to execute commands.

File: Justfile [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+             else { "Note: 'Too Many Requests' errors are allowed here but not in CI, set GITHUB_TOKEN to check them" } }}
+
+# build target/rpc_errors_schema.json
+build-rpc-errors-schema:
+    rm -f target/rpc_errors_schema.json
+    cargo check -p near-jsonrpc --features dump_errors_schema
+
+# update chain/jsonrpc/res/rpc_errors_schema.json
+update-rpc-errors-schema: build-rpc-errors-schema
+    cp target/rpc_errors_schema.json chain/jsonrpc/res/rpc_errors_schema.json
```
Comment (nagisa) [inline Justfile:80]: Any reason these have been split up into two?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -56,34 +54,28 @@ jobs:
...
           flags: unittests,${{ matrix.id }}
       # See https://github.com/taiki-e/cargo-llvm-cov/issues/292
       - run: find target -name '*.profraw' -delete
 
       # Run integration tests
-      - run: cargo nextest run --locked --package integration-tests --cargo-profile dev-release --profile ci ${{ matrix.flags }}
-        if: matrix.run_integ_tests
+      - run: just codecov "nextest-integration ${{ matrix.type }}"
         env:
           RUST_BACKTRACE: short
```
Comment (nagisa) [inline .github/workflows/ci.yml:74]: Observation, with `env_var_or_default` we could move this into the Justfile with something along the lines of

```just
export RUST_BACKTRACE := get_env_or_default("RUST_BACKTRACE", "short")
```

or

```just
rust_backtrace := get_env_or_default("RUST_BACKTRACE", "short")

nextest_integration: $RUST_BACKTRACE="{{rust_backtrace}}"
    ...
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -56,34 +54,28 @@ jobs:
...
-      - run: cargo llvm-cov report --profile dev-release --codecov --output-path unittests.json
       - uses: codecov/codecov-action@eaaf4bedf32dbdc6b720b63067d99c4d77d6047d
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
-          files: unittests.json
+          files: codecov.json
           fail_ci_if_error: true
           flags: unittests,${{ matrix.id }}
       # See https://github.com/taiki-e/cargo-llvm-cov/issues/292
       - run: find target -name '*.profraw' -delete
```
Comment (nagisa) [inline .github/workflows/ci.yml:69]: Should this not be a part of the Justfile as well? And then perhaps check if `--no-clean` is passed to `just` and disable the this delete there?

File: Justfile [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+# generate a codecov report for RULE
+codecov RULE:
+    #!{{ bash }}
+    set -euxo pipefail
+    source <(cargo llvm-cov show-env --export-prefix)
+    just {{ RULE }}
+    cargo llvm-cov report --profile dev-release --codecov --output-path codecov.json
+
+# style checks
+check-style:
```
Comment (nagisa) [inline Justfile:61]:
```suggestion
check-python-style:
```

File: Justfile [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    python3 scripts/check_nightly.py
+    python3 scripts/check_pytests.py
+    python3 scripts/fix_nightly_feature_flags.py
+    ./scripts/formatting --check
+
+# lychee-based url validity checks
+check-lychee:
+    lychee . {{ if env("GITHUB_TOKEN", "") != "" { "" } else { "-a 429" } }}
```
Comment (nagisa) [inline Justfile:69]: This is not getting run in GHA though, is it? Perhaps consider throwing an `error()` for GHA environment instead, to make it clear to the future readers that this is a local-only step.

**Review by nagisa — COMMENTED** [review]
No hurry, also happy to abandon my claim to each individual point.

File: Justfile [hunk 9]
```diff
@@ -0,0 +1,84 @@
+# FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
+# them at earliest convenience :)
+# Also in addition to this, the `nextest-integration` test is currently disabled on macos
+with_macos_excludes := if os() == "macos" {
+    "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+} else {
+    ""
+}
+nightly_flags := "--features nightly,test_features"
+bash := if os() == "macos" { "/bin/bash" } else { "/usr/bin/bash" }
```
[the remark this answers is not in the document]
Comment (Ekleog-NEAR) [reply Justfile]: Great point! I hadn’t thought that env was available at the same place on linux and macos runners :)

File: Justfile (hunk 1 above)
Comment (Ekleog-NEAR) [reply Justfile:21]: Hmm… I do agree that it would be good to have this, but I’m not sure `Justfile` would be the right tool for it? Basically, this would be going away from the behaviour of CI, and so is probably less important.

And TBH I don’t think it makes sense for people to learn the syntax for our specific `Justfile` when they already know the syntax for `cargo`. So I’m thinking of `just` as just a way to reproduce CI behavior locally, and nothing more.

WDYT?

File: Justfile (hunk 2 above)
Comment (Ekleog-NEAR) [reply Justfile:30]: Hmm… so similarly, this is arguably moving away from the behaviour of CI. So I’m not sure it makes sense to remove it from the `Justfile`, if the goal of `just` is to reproduce CI behaviour locally?

File: Justfile (hunk 3 above)
Comment (Ekleog-NEAR) [reply Justfile]: Right; I’m thinking of `just` as basically a way to multiplex bash scripts easily, where it’s not too bad to have that kind of recursion. Basically, it doesn’t seem to support easily wrapping commands around another command, because it’s still thought as "similar to make" I guess, and make wouldn’t support that. It’d maybe be better if it were a generic scripting language with support for multiple commands, but…

File: Justfile (hunk 4 above)
Comment (Ekleog-NEAR) [reply Justfile:81]: This way local development can use `update` and CI `check`, without duplicating the code in`build` :)

File: .github/workflows/ci.yml (hunk 5 above)
Comment (Ekleog-NEAR) [reply .github/workflows/ci.yml:74]: Hmm so I was actually wondering, what’s the point of setting the backtrace to `short` in CI? I remember digging through CI and not finding enough details to pinpoint the failure, so I’m thinking CI should have at least a full backtrace, though local development could also work with short backtraces.

So I’m curious, should we not just make it a `RUST_BACKTRACE=1` alongside moving it to the Justfile?

File: .github/workflows/ci.yml (hunk 6 above)
Comment (Ekleog-NEAR) [reply .github/workflows/ci.yml:69]: Hmm so it’s using the codecov token, which allows uploading coverage reports for any PR. So it’d be bad to make it public. From that on, I don’t see the point in making this part of the Justfile.

File: Justfile (hunk 8 above)
Comment (Ekleog-NEAR) [reply Justfile:70]: It’s not running in GHA. But it’s actually useful to have `GITHUB_TOKEN` set locally, if you want to check URL validity locally: github rate limits are such that the second run of `lychee` already hits rate limits. I actually had to have one in order to make the first PR to fix all our dead links :)

File: .github/workflows/ci.yml (hunk 5 above)
Comment (nagisa) [reply .github/workflows/ci.yml:74]: My opinion is that the fluff that `short` filters out is very, very rarely useful in programs that don’t deal with some particularly low level details (which neard does not.)

File: Justfile (hunk 4 above)
Comment (nagisa) [reply Justfile:81]: I don’t personally find (check not updating the file that is being checked) a useful thing to do, but alright.

Actually, I feel that it would still be better to check with `git diff` as it was done before, as ultimately what we care about is “is chain/jsonrpc/res/rpc_errors_schema.json the same as the checked-in file” and not “is the newly built file equal to whatever file is currently checked out (including any modifications not checked in)”.

But no strong opinion here.

Comment (Ekleog-NEAR) [reply Justfile:81]: Hmm I see your point, I guess my thinking is mostly "`just test` should not affect the worktree outside of `target`, as that’d be quite unexpected" ^^’

That said, as you say, I think we can reasonably live with both options :)

File: .github/workflows/ci.yml (hunk 6 above)
Comment (Ekleog-NEAR) [reply .github/workflows/ci.yml:69]: Oooooh yes ok I get it, I thought it was for the `codecov-action` 😅 This is now implemented!

File: Justfile (hunk 2 above)
Comment (nagisa) [reply Justfile:32]: `--profile ci` is *the* hack here, necessitated only because we have spuriously failing tests that were a significant hindrance on the development velocity when CI would get stuck on them. It isn’t an ideal place to be in the first place, but we cannot really afford the merge queue kicking out PRs occasionally when a test unrelated to those PRs decides to explode.

OTOH `--profile ci` is a reason why `ed25519-dalek` upgrade required so much time to work out – the spurious failure introduced by an earlier change was hidden by the retries enabled by `--profile ci`.

Despite that I still feel that having a couple retries on a per-test basis in CI *only* is the least terrible of the options available to us given the current requirements and the state of our test suite.

I believe this is one of those things that we will have to live with being different.

File: Justfile (hunk 1 above)
Comment (nagisa) [reply Justfile:23]: > And TBH I don’t think it makes sense for people to learn the syntax for our specific Justfile when they already know the syntax for cargo. So I’m thinking of just as just a way to reproduce CI behavior locally, and nothing more.

How does this align with the intent to move the `style` checks out from within nextest and the hope that we eventually have a single command local development experience, not multiple?

Like, if Justfile’s singular concern is reproducing the CI, then it is not all that appealing in majority of local development workflows (e.g. I’d imagine rarely anybody runs the entire test suite in every local iteration cycle) which then makes it a poor replacement for the current `style` tests in nextest too…?

File: Justfile (hunk 2 above)
Comment (Ekleog-NEAR) [reply Justfile:32]: Makes sense, I added it as an environment-configured toggle :)

File: Justfile (hunk 1 above)
Comment (Ekleog-NEAR) [reply Justfile:23]: Hmm… I don’t think the current `style` tests are any better than a `just check-style` command would be? If being run with `-p near-vm-runner`, then the style tests still won’t be run.

In my mind it’s a balance to find between "re-implementing the world behind a new CLI" (the solution of offering toggles for everything, though a reasonable amount of toggles, like a `CARGOFLAGS` environment variable, might make sense) and "having a single command to run all of CI in a monolithic way".

So basically my goal here is to have many commands to run each part of CI, more or less configurable depending on what’s easily doable, and then people can mix-and-match them depending on their needs. TBH I think I personally will mostly want to just run the `just` command that’s being run by whichever CI job I see failing, and just use regular `cargo nextest -p whichevercrateImworkingon` usually 😅 

WDYT?

Comment (Ekleog-NEAR) [reply Justfile:24]: Hmm actually… maybe just having `just packages=near-vm-runner,node-runtime test` run all the test suite (including style checks) but only with `-p near-vm-runtime -p node-runtime` would be enough to address your concerns?

It’d be strictly better than the statu quo, where style checks don’t run when nextest is run with `-p`, and it wouldn’t require too large an additional API surface :)

WDYT?

Comment (Ekleog-NEAR) [reply Justfile]: Good point! This is now done, though using `just nextest stable -p near-vm-runner` will have the maybe-unexpected result of also running tests for the `integration-tests` package :)

Comment (Ekleog-NEAR) [reply Justfile]: I didn’t extend it to all of `just test`, because I think there’d be a risk of misunderstanding to which sub-command the additional flags would apply; so it currently applies only to `nextest`, `nextest-unit` and `nextest-integration`.

**Review by nagisa — APPROVED** [review]
the `nextest` (the plain one) still specifies `--profile ci`, but otherwise LGTM.

I think we can iterate on this from here on.

**Comment by Ekleog-NEAR** [comment]
Nice catch, thank you! This is now fixed :)
